yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -903,6 +908,30 @@ object Client extends Logging {
   }
 
   /**
+   * Obtains token for the Hive metastore and adds them to the credentials.
+   */
+  private def obtainTokenForHiveMetastore(conf: Configuration, credentials: Credentials) {
+    if (UserGroupInformation.isSecurityEnabled /* And Hive is enabled */) {
+      val hc = org.apache.hadoop.hive.ql.metadata.Hive.get
+      val principal = hc.getConf().get(HiveConf.ConfVars.METASTORE_KERBEROS_PRINCIPAL.varname)
+      val username = UserGroupInformation.getCurrentUser().getUserName
+
+      if (principal == null) {
+        val errorMessage = "Required hive metastore principal is not configured!"
+        logError(errorMessage)
+        throw new IllegalArgumentException(errorMessage)
+      }
+
+      val tokenStr = hc.getDelegationToken(username,principal)
+      val hive2Token = new Token[DelegationTokenIdentifier]()
+      hive2Token.decodeFromUrlString(tokenStr)
+      credentials.addToken(new Text("hive.server2.delegation.token"), hive2Token)
+      logDebug("Added the Hive Server 2 token to conf.")
+      org.apache.hadoop.hive.ql.metadata.Hive.closeCurrent
```

Comment:
Hbase needs its own delegation token, so I guess it wouldn't work either without some code to fetch the token.

Regarding token expiry, especially for a long running process (let's say the thrift server), it would require something similar to what Hari is doing in #4688 .